ports/nrf/modules/machine/uart.c

```diff
@@ -323,27 +338,36 @@ STATIC MP_DEFINE_CONST_DICT(machine_hard_uart_locals_dict, machine_hard_uart_loc
 STATIC mp_uint_t machine_hard_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
     const machine_hard_uart_obj_t *self = self_in;
     byte *buf = buf_in;
+    uint32_t t = self->timeout + mp_hal_ticks_ms();
 
     // read the data
     for (size_t i = 0; i < size; i++) {
         while (!uart_rx_any(self)) {
+            if (mp_hal_ticks_ms() > t) {  // timed out
+                if (i <= 0) {
+                    *errcode = MP_EAGAIN;
+                    return MP_STREAM_ERROR;
+                } else {
+                    return i;
+                }
+            }
+            MICROPY_EVENT_POLL_HOOK
         }
         buf[i] = uart_rx_char(self);
+        t = self->timeout_char + mp_hal_ticks_ms();
     }
 
     return size;
 }
 
 STATIC mp_uint_t machine_hard_uart_write(mp_obj_t self_in, const void *buf_in, mp_uint_t size, int *errcode) {
     machine_hard_uart_obj_t *self = self_in;
-    const byte *buf = buf_in;
-
-    nrfx_err_t err = NRFX_SUCCESS;
-    for (int i = 0; i < size; i++) {
-        err = uart_tx_char(self, (int)((uint8_t *)buf)[i]);
-    }
 
+    nrfx_err_t err = nrfx_uart_tx(self->p_uart, buf_in, size);
     if (err == NRFX_SUCCESS) {
+        while (machine_uart_txdone(self) == mp_const_false) {
```

Comment:
I think it's clearer to just call `nrfx_uart_tx_in_progress()` directly here.  Or make a helper like `machine_uart_tx_in_progress()` that returns a `bool` and use it here and in `machine_uart_txdone()`.

Reply:
I deleted that loop, because it caused uart.write() to wait until the transmission was finished, making flush() and txdone() kind of obsolete.

Reply:
No, that loop is required. So I moved it up before the send to wait until a previous transmit is finished.

Reply:
Can't move it up, since the data is not internally buffered. So uart.write() waits until all data has been sent, and txdone() could simply return True, and flush could return w/o testing.

Reply:
I see you made the implementation of txdone and flush still check the buffer (instead of doing nothing).  I guess that's sensible, in case things improve/change in the future with this driver and buffering.

Reply:
I removed these checks, and inserted them again. They are short, and having them in place is good for future extension. Also, with the MICROPY_EVENT_POLL_HOOK in the wait loop of uart_write a scheduled task could run run write, which may check for UART tx being finished, even if that doesn't make sense.

Reply:
PR updated & force-pushed.